### PR TITLE
Add possibility to initialise examples from cli

### LINKED ISSUE #1339: Make it possible to initialize examples from cli
Currently there are some `ert3` examples in `examples/`. It would be neat for developer and user testing to be able to fetch these directly from the command line tool itself. This could for instance be done by adding an optional `--example` flag to `ert3 init` that would accept any of the workspace names residing in `examples/`. Then, one could write `ert3 init --example polynomial` and have a folder named `polynomial` appear containing the content of `examples/polynomial`.

The command should probably fail if the folder `polynomial` already exists in the example usage of example initialization above...

### PR DESCRIPTION
**Pull request by kvashchuka**
**Issue**
Resolves #1339 

**Approach**
Optional flag `--example` was added, such that when you run `ert3 init --example EXAMPLE` EXAMPLE is copied to your working directory and initialised. EXAMPLE is copied from `examples` in virtual environment as `examples` are now distributed together with ert

### CODE REVIEW

**Comment by markusdregi** [comment]
This seems to be heading in a good direction 👍 In particular, `--example` should be optional. If provided, the valid values are the names of the different workspaces in `examples`. Given `ert3 init --examples polynomial` I expect it to fail if there already is a folder in cwd named polynomial. Otherwise, I expect it to create a folder named polynomial with the content in `examples/polynomial`. Notice, that this should work from the installed version without the GitHub repository, which implies that the examples will have to be bundled with the `ert3` module.

**Comment by sondreso** [comment]
> Notice, that this should work from the installed version without the GitHub repository, which implies that the examples will have to be bundled with the ert3 module.

Or fetch them from the ERT repository on demand?

**Comment by sondreso** [comment]
Fetching a sub-directory from a git repo is not straight forward, but it might be worth checking out if we plan to have large examples. Options include using some 3rd-party service (see: https://stackoverflow.com/questions/7106012/download-a-single-folder-or-directory-from-a-github-repo) or maybe use the github api directly?

**Comment by kvashchuka** [comment]
yes, thank you @sondreso ! That's something I wanted to discuss with you

File: tests/ert3/console/integration/test_cli.py [hunk 1]
```diff
@@ -52,6 +52,29 @@ def test_cli_init(tmpdir):
...
+def test_cli_init_example(tmpdir):
+    workspace = tmpdir / _POLY_WORKSPACE_NAME
+    shutil.copytree(_POLY_WORKSPACE, workspace)
+    workspace.chdir()
+
+    args = ["ert3", "init", "--example", "something"]
+    with patch.object(sys, "argv", args):
+        with pytest.raises(SystemExit) as error:
+            ert3.console.main()
+        assert "There is no example something in ert/examples." in str(error.value)
```
Comment (markusdregi) [inline tests/ert3/console/integration/test_cli.py:64]: This checking can be done with the optional `match` argument to `raises` :)

File: tests/ert3/console/integration/test_cli.py [hunk 2]
```diff
@@ -52,6 +52,29 @@ def test_cli_init(tmpdir):
...
 
 
+def test_cli_init_example(tmpdir):
+    workspace = tmpdir / _POLY_WORKSPACE_NAME
+    shutil.copytree(_POLY_WORKSPACE, workspace)
+    workspace.chdir()
+
+    args = ["ert3", "init", "--example", "something"]
+    with patch.object(sys, "argv", args):
+        with pytest.raises(SystemExit) as error:
```
Comment (markusdregi) [inline tests/ert3/console/integration/test_cli.py:62]: Should this not be an `IllegalWorkspaceOperation`?

**Review by markusdregi — COMMENTED** [review]
Just some early and minor comments :)

**Comment by kvashchuka** [comment]
@markusdregi when you proposed `ert3 init --example example_name` did you assume that the folder will be copied and the copied example will be initialised or not? Cause if not than it probably should not be called `init`

**Comment by markusdregi** [comment]
@kvashchuka Yes, with the command I would expect the example to be persisted to disk and be initialised 👍

File: tests/ert3/console/integration/test_cli.py (hunk 2 above)
Comment (kvashchuka) [reply tests/ert3/console/integration/test_cli.py:62]: I don't think it would work cause in main we catch `sys.exit(e)` or am I wrong and it doesn't matter?

```
def main():
    try:
        _main()
    except ert3.exceptions.ErtError as e:
        sys.exit(e)
```

Comment (markusdregi) [reply tests/ert3/console/integration/test_cli.py:62]: You are correct 👍

File: tests/ert3/console/integration/test_cli.py [hunk 3]
```diff
@@ -52,6 +52,29 @@ def test_cli_init(tmpdir):
...
+        assert "Example something is not a valid ert3 example." in str(error.value)
+
+    args = ["ert3", "init", "--example", "polynomial"]
+    with patch.object(sys, "argv", args):
+        ert3.console.main()
+
+    with patch.object(sys, "argv", args):
+        with pytest.raises(SystemExit) as error:
+            ert3.console.main()
+        assert "Your working directory already contains example polynomial" in str(
```
Comment (markusdregi) [inline tests/ert3/console/integration/test_cli.py:73]: The `match` argument to `pytest.raises` is made for this intention 👍

**Comment by markusdregi** [comment]
Would be nice with a CI pipeline that initialises the polynomial example using the example option, to ensure that this works in full integration test on the installed version.

We can get this either by adding an additional pipeline, or by changing the current example pipeline to not use the example in the repository, but the `--example` flag instead. Maybe the later would be neat 👏

File: ert3/console/_console.py [hunk 4]
```diff
@@ -52,14 +66,41 @@ def _build_argparser():
...
+
+    if args.example is None:
+        ert3.workspace.initialize(Path.cwd())
+    else:
+        _EXAMPLE_NAME = args.example
+        pkg_examples_path = pathlib.Path(pkg.resource_filename("ert", "../examples"))
+        pkg_example_path = pkg_examples_path / _EXAMPLE_NAME
+        wd_example_path = Path.cwd() / _EXAMPLE_NAME
+        if not pkg_examples_path.exists():
+            raise ert3.exceptions.IllegalWorkspaceOperation(
```
Comment (jondequinor) [inline ert3/console/_console.py:88]: Maybe include information about what path we tried to load?

File: ert3/console/_console.py [hunk 5]
```diff
@@ -52,14 +66,41 @@ def _build_argparser():
...
+                "Examples folder was not found."
+            )
+        if not pkg_example_path.exists():
```
Comment (jondequinor) [inline ert3/console/_console.py:91]: Same here, maybe include information about what path was checked.

File: examples/polynomial/run_demo [hunk 6]
```diff
@@ -1,7 +1,8 @@
 #!/usr/bin/env bash
 set -e
 
-ert3 init
+ert3 init --example polynomial
```
Comment (markusdregi) [inline examples/polynomial/run_demo:4]: This is a bit odd, as now we create a new workspace which is a copy of what was the intended workspace within the workspace 🤷 If we go down this route, which is probably a nice idea, we should not change the run_demo script, but the CI-pipeline. So that instead of it moving into the correct workspace and run poly_demo it does:

```
ert3 init --example polynomial (or spe1)
pushd polynomial
./run_demo
popd
```
and then we remove the init command from the `run_demo` script.

File: tests/ert3/console/integration/test_cli.py [hunk 7]
```diff
@@ -35,6 +35,27 @@ def test_cli_init(tmpdir):
             ert3.console.main()
 
 
+def test_cli_init_example(tmpdir):
```
Comment (markusdregi) [inline tests/ert3/console/integration/test_cli.py:38]: Feel free to split this up into three tests named according to what they test

File: tests/ert3/console/integration/test_cli.py [hunk 8]
```diff
@@ -35,6 +35,27 @@ def test_cli_init(tmpdir):
...
+        args = ["ert3", "init", "--example", "something"]
+        with patch.object(sys, "argv", args):
+            with pytest.raises(
+                SystemExit, match="Example something is not a valid ert3 example."
+            ):
+                ert3.console.main()
+
+        args = ["ert3", "init", "--example", "polynomial"]
+        with patch.object(sys, "argv", args):
+            ert3.console.main()
```
Comment (markusdregi) [inline tests/ert3/console/integration/test_cli.py:49]: Should we do any checking that we actually created some files in this example?

File: ert3/console/_console.py [hunk 9]
```diff
@@ -1,17 +1,30 @@
...
     "all users to ert for real work!"
 )
 
 
+def _build_init_argparser(subparsers):
+    init_parser = subparsers.add_parser("init", help="Initialize an ERT3 workspace")
+    init_parser.add_argument(
+        "--example",
+        help="Name of the example that would be copied "
+        "to the working directory and initialised",
```
Comment (markusdregi) [inline ert3/console/_console.py:24]: It would be really neat to list the available examples here. If you choose to not address it in this PR, it would be nice if you could create an issue and write in the PR description that you defer that to later efforts :)

**Review by markusdregi — COMMENTED** [review]
Did you look into other alternatives for bundling the examples such that they become resources in the `ert3` module? As of now we instead create a new Python module named `examples`... Probably not how Python intends us to resolve this 🤔 How about options like `package_data` in `setup.py`?

**Comment by markusdregi** [comment]
🎉 Happy news 🎉 

I can confirm that this works exactly as I hoped! I just cloned your branch, installed it and then moved into a completely different folder and did:

```
ert3 init --example polynomial
cd polynomial
ert3 run evaluation
```

🚀

File: examples/polynomial/run_demo (hunk 6 above)
Comment (mortalisk) [reply examples/polynomial/run_demo:4]: Only weird if you run the script within the workspace, which you do not have to do I guess. Perhaps moving the run_demo script to e.g. tests/ert3demos/run_polynomial_demo would make it clear that it is supposed to work from anywhere.

Comment (markusdregi) [reply examples/polynomial/run_demo:4]: Sure, but then it is no longer a demo script... Currently, the demo script is part of the example in the sense that you can run it and get some results and inspect them afterwards. It also serves as an example and guidance for how one can interact with an ert3 workspace. We can move it into `tests/` and rename it to `test_polynomial`, but then we also need to accept that the examples does not come with a demo script. For which I'm fine with if that is your argument ;)

Comment (mortalisk) [reply examples/polynomial/run_demo:4]: Mostly wanted to point out that it doesn't need to be run from the examples folder. I think having a demo makes sense.

Comment (markusdregi) [reply examples/polynomial/run_demo:4]: Sure, but that is where it currently resides and then it is a bit odd that it recreates the data (and a new demo script) within the workspace again :)

Comment (mortalisk) [reply examples/polynomial/run_demo:4]: I agree. Another argument in favor of your position is that a regular user, in order to run the demo, would presumably have to run init simply to get the demo script anyways, so the init as part of the demo is just confusing.

File: ert3/console/_console.py (hunk 9 above)
Comment (kvashchuka) [reply ert3/console/_console.py:24]: Make sense 👍

File: ert3/console/_console.py [hunk 10]
```diff
@@ -1,17 +1,30 @@
-import ert3
-
 import argparse
-from pathlib import Path
+import pathlib
```
Comment (mortalisk) [inline ert3/console/_console.py:2]: This import seems unnecessary as only the Path symbol is used, which is also imported below.

File: ert3/console/_console.py [hunk 11]
```diff
@@ -77,6 +90,37 @@ def _build_argparser():
     return parser
 
 
+def _init(args):
+    assert args.sub_cmd == "init"
+
+    if args.example is None:
+        ert3.workspace.initialize(Path.cwd())
+    else:
+        _EXAMPLE_NAME = args.example
```
Comment (mortalisk) [inline ert3/console/_console.py:99]: This could be named simply example_name. Underscore indicates it is private, but all locals are private. Capital letters indicate it is a shared constant, but it is a local to only this function.

**Comment by kvashchuka** [comment]
> Did you look into other alternatives for bundling the examples such that they become resources in the `ert3` module? As of now we instead create a new Python module named `examples`... Probably not how Python intends us to resolve this 🤔 How about options like `package_data` in `setup.py`?

Yes, of course, I started with `package_data` and MANIFEST.in file, but I could not make it work.. But maybe I was doing it wrong

**Comment by markusdregi** [comment]
Hmm 🤔 A long time since I used package data... Did you try having a look at it with someone else from the team? Perhaps four eyes are needed :)

File: ert3/console/_console.py (hunk 11 above)
Comment (kvashchuka) [reply ert3/console/_console.py:99]: my bad

File: ert3/console/_console.py (hunk 9 above)
Comment (kvashchuka) [reply ert3/console/_console.py]: I decided to add list of available examples in this PR

**Comment by kvashchuka** [comment]
> Would be nice with a CI pipeline that initialises the polynomial example using the example option, to ensure that this works in full integration test on the installed version.
> 
> We can get this either by adding an additional pipeline, or by changing the current example pipeline to not use the example in the repository, but the `--example` flag instead. Maybe the later would be neat 👏

This might be a bad idea given that now I am adding `examples` to ert distribution, which means that `polynimal` and `spe1` examples version will be the same as in the latest release of ert. Which means that changes in `examples` in future PRs will not be tested by `run demo` workflows
And this brings us to the question whether I need to include it in ert distribution, cause as far as I understood, usually wheel distribution should not contain examples/data ...

**Comment by markusdregi** [comment]
> This might be a bad idea given that now I am adding examples to ert distribution, which means that polynimal and spe1 examples version will be the same as in the latest release of ert. Which means that changes in examples in future PRs will not be tested by run demo workflows

I don't fully understand 🤔 Because when testing we are always using the relevant version of ERT, which would then give the same version of the examples no?

> And this brings us to the question whether I need to include it in ert distribution, cause as far as I understood, usually wheel distribution should not contain examples/data ...

It is not a requirement to add it to the package, but I believe that is the easiest way to resolve this... If the size of the package becomes problematic I think we can deal with that later.

**Comment by markusdregi** [comment]
I was kind of hoping that we could keep the examples folder where it is, and use something like `package_data` to point to it. But maybe it does not support data outside of the folder of the module? Perhaps we could have `setup.py` copy if over at installation?

Either way I think this `ert3.examples` module is a good place to distribute them 👍🏻

**Comment by kvashchuka** [comment]
> I was kind of hoping that we could keep the examples folder where it is, and use something like `package_data` to point to it. But maybe it does not support data outside of the folder of the module? Perhaps we could have `setup.py` copy if over at installation?
> 
> Either way I think this `ert3.examples` module is a good place to distribute them 👍🏻

I can leave the folder `examples` in `ert` and rename it `ert3_examples` and distribute as a package for example. The thing is: `package_data` can only add files to existing packages, so we either ship examples as a package separately (and rename it), or as part of `ert3` (and move the folder there)

**Comment by markusdregi** [comment]
I see, but you cannot introduce the module `ert3.examples` with only an init-file and then add `examples` from the root-directory as `package_data` to that submodule? That would be the best option, but if not possible I suggest we rename `examples` to `ert3_examples` and learn to live with the `__init__` file inside it.

**Comment by markusdregi** [comment]
Jenkins: Test this please

File: tests/ert3/console/integration/test_cli.py [hunk 12]
```diff
@@ -63,6 +63,44 @@ def test_cli_init_subfolder(workspace):
...
+
+@pytest.mark.requires_ert_storage
+def test_cli_init_example(tmpdir, ert_storage):
+    with tmpdir.as_cwd():
+        args = ["ert3", "init", "--example", "polynomial"]
+        with patch.object(sys, "argv", args):
+            ert3.console.main()
+        polynomial_path = tmpdir / "polynomial"
+        if polynomial_path.exists() and polynomial_path.isdir():
+            # check that copied polynomial example is not an empty directory
```
Comment (markusdregi) [inline tests/ert3/console/integration/test_cli.py:85]: You can probably assert that `stages.yml`, `parameters.yml` and `experiments` also is inside the folder. If not, it does not look like a workspace :)

File: tests/ert3/console/integration/test_cli.py [hunk 13]
```diff
@@ -63,6 +63,44 @@ def test_cli_init_subfolder(workspace):
...
+            assert len(polynomial_path.listdir()) > 0
+
+
+@pytest.mark.requires_ert_storage
+def test_cli_init_example_twice(tmpdir, ert_storage):
```
Comment (markusdregi) [inline tests/ert3/console/integration/test_cli.py:90]: Could you add a similar test where one creates a workspace, moves into it and then tries to create a second workspace?

File: ert3/console/_console.py [hunk 14]
```diff
@@ -78,6 +96,44 @@ def _build_argparser():
     return parser
 
 
+def _get_ert3_examples_path():
+    pkg_examples_path = pathlib.Path(pkg.resource_filename("ert", "../ert3_examples"))
```
Comment (markusdregi) [inline ert3/console/_console.py:100]: Would it not be more correct to ask for `pkg.resorce_filename("ert3_examples")`?

File: .github/workflows/run_examples_spe1.yml [hunk 15]
```diff
@@ -72,7 +72,9 @@ jobs:
 
     - name: Run spe1 demo
       run: |
-        pushd examples/spe1
+        ert3 init --example spe1
+        pushd spe1
         ./run_demo
         popd
 
+
```
Comment (markusdregi) [inline .github/workflows/run_examples_spe1.yml:80]: Unnecessary whitespace...

File: MANIFEST.in [hunk 16]
```diff
@@ -0,0 +1,2 @@
+include ert_gui/resources/*
+include ert_logging/*
```
Comment (markusdregi) [inline MANIFEST.in:2]: Missing whitespace

File: pyproject.toml [hunk 17]
```diff
@@ -0,0 +1,2 @@
+[build-system]
+requires = ["setuptools", "wheel", "setuptools_scm"]
```
Comment (markusdregi) [inline pyproject.toml:2]: Missing whitespace

**Review by markusdregi — COMMENTED** [review]
This is looking good 👍🏻 Can you rebase your history and make a few commits out of it. Especially, it would be nice if you get the movement of all the files into a single commit...

Also, could you explain why we are replacing `package_data` with a `MANIFEST`? :)

**Comment by kvashchuka** [comment]
> This is looking good 👍🏻 Can you rebase your history and make a few commits out of it. Especially, it would be nice if you get the movement of all the files into a single commit...
> 
> Also, could you explain why we are replacing `package_data` with a `MANIFEST`? :)

Øyvind @oyvindeide suggested that this is a more modern way to do it. Plus, we should not use both `package_data` and a `MANIFEST`. However, when I made a change I was planning to use `MANIFEST`, but we ended up with a separate module `et3_examples`. So to be honest, that is not part of the scope anymore, and I can leave the things as they were before. Should I revert this change?

**Comment by markusdregi** [comment]
If it is not necessary I suggest you revert and either bring up the discussion in an issue or create an isolated PR with those changes alone 👍🏻

File: tests/ert3/console/integration/test_cli.py (hunk 13 above)
Comment (kvashchuka) [reply tests/ert3/console/integration/test_cli.py:90]: Currently if you do 
```
ert3 init --example spe1
cd spe1
ert3 init --example spe1
```
spe1 example will be copied inside spe1 but won't be initialised and you will see an error "Already inside an ert workspace." So this behavior is covered by `test_cli_init_subfolder` test

I can check whether we are already inside the folder with the same name as provided example name and that it has attributes of example (like stages.yml etc.), but I don't think I can check whether all files in the current working directory are identical to the one I am about to copy, and I am not sure we need that... or we do?

Comment (markusdregi) [reply tests/ert3/console/integration/test_cli.py:90]: I’m thinking the natural behavior would be to detect that it is already within a workspace and neither copy the files nor initialize. Otherwise, we carry out half the operation and then abandon 🤷🏻

**Comment by oyvindeide** [comment]
This might be slightly besides the point, and feel free to ignore it - but we might want to reconsider how we bundle `ert` 2.x and `ert3`, because at the moment if someone does `pip install ert==2.21` they also get `ert3`, which bundles its own executable and probably does not have a stable API. As we strive to maintain semver in our projects, and also because we would like to keep our users happy, we should perhaps not include `ert3` in the distributed version, or at the very least write a disclaimer in the `README` as there is no mention of it today.

And if you are testing something experimental, we should perhaps not include all kinds of convenience (like examples) just for the sake of it? It should be fine to get it from GitHub, sure, it is one more step, but it is the _right_ thing do do - I think

**Comment by markusdregi** [comment]
Good discussion topic @oyvindeide! But I don't think it belongs in this PR ;) To give a very short summary `ert3 --help` gives you a pretty fat disclaimer of it being experimental. Any documentation of `ert3` is yet to be written and I think we should sort out some of the directional choices first, before we spend too much time on it. We did look into options for making installation of `ert3` optional earlier, but did not find an immediate solution so it was left for later discussions... As the cli states that it is an experimental tool, I think it is clear that semver does not apply. If you want to continue this discussion or have concrete suggestions I suggest we move it to an issue or PR on this very topic 👍🏻

**Comment by markusdregi** [comment]
When it comes to being able to initialise from an example, this makes user testing a lot easier. And also the threshold for testing out the tool in the future quite easy. Also, since it is indeed moving fast it might be non-trivial to find the correct version of the examples compared to the version you have installed. The cli-option takes care of that.

The motivation was and is to make user testing as easy as possible and that I believe is worth quite a bit! The PR itself ended up being a bit more substantial then I did foresee, but hopefully it was a valuable learning experience for @kvashchuka on bundling data into a Python distribution 🤷🏻 I think we should strive to get this into a mergeable state as soon as possible :)

**Comment by markusdregi** [comment]
Jenkins: Test this please

**Comment by oyvindeide** [comment]
> Good discussion topic @oyvindeide! But I don't think it belongs in this PR ;) 

No, I agree with that, but there seemed to be more discussion here than in the corresponding issue but as I mentioned, feel free to disregard my comments ;)

>To give a very short summary `ert3 --help` gives you a pretty fat disclaimer of it being experimental. 

Sure, but there is nothing preventing me from `from ert3.engine import x` without any warning, so the executable is documented, but the actual code is not.

> We did look into options for making installation of `ert3` optional earlier, but did not find an immediate solution so it was left for later discussions... 

Yes, python does not really support that natively, and I see the value of having it in the same repository, where we are going to come across these problems.

> When it comes to being able to initialise from an example, this makes user testing a lot easier. And also the threshold for testing out the tool in the future quite easy. Also, since it is indeed moving fast it might be non-trivial to find the correct version of the examples compared to the version you have installed. The cli-option takes care of that.

Sure, but so does installing directly from GitHub, and you could make a source distribution including the examples, that would keep the wheel clean from stuff that is not really meant to be there by the python overlords.

File: ert3/console/_console.py (hunk 14 above)
Comment (kvashchuka) [reply ert3/console/_console.py:100]: `resource_filename` needs two arguments and it wouldn’t work with just `ert3_examples`. However, I do agree that it looks weird, maybe I am misusing function `resource_filename` and it should be done differently

Comment (markusdregi) [reply ert3/console/_console.py:100]: I think what is weird is that we are using the location of `ert` to locate `ert3_examples`. And this is not necessarily correct, although it is as installation is set up today...

I think `pkg.resource_filename("ert3_examples", ".")` is better and should do the same for you 🤔

**Comment by markusdregi** [comment]
Sure @oyvindeide, but then testing on any version that we distribute get much more cumbersome... I think providing examples easily is a really neat feature and if we take responsibility for providing the correct version of the example that gives us the opportunity to change the distribution mechanism of them completely down the road.

This PR is almost done, provides a really neat feature and hence I'm voting for merging it ASAP. Alternative implementations are most definitively welcome afterwards 👍🏻

File: ert3/console/_console.py [hunk 18]
```diff
@@ -13,6 +16,21 @@
 )
 
 
+def _build_init_argparser(subparsers):
+    init_parser = subparsers.add_parser("init", help="Initialize an ERT3 workspace")
+    pkg_examples_path = _get_ert3_examples_path()
+    ert_example_names = []
```
Comment (markusdregi) [inline ert3/console/_console.py:22]: This looks like a good candidate for a helper function in the sense of `_get_ert3_example_names`.

File: ert3/console/_console.py [hunk 19]
```diff
@@ -78,6 +96,44 @@ def _build_argparser():
     return parser
 
 
+def _get_ert3_examples_path():
+    pkg_examples_path = pathlib.Path(pkg.resource_filename("ert3_examples", ""))
+    # check that examples folder exist
+    if not pkg_examples_path.exists():
+        raise ert3.exceptions.IllegalWorkspaceOperation(
```
Comment (markusdregi) [inline ert3/console/_console.py:103]: This seems to be an internal error...? Perhaps it should not be an `ErtError`, but present itself as a stack trace and a bug?

Comment (markusdregi) [reply ert3/console/_console.py:103]: Also, adding `pkg_examples_path` to the error message will help debugging...

File: ert3/console/_console.py [hunk 20]
```diff
@@ -78,6 +96,44 @@ def _build_argparser():
...
+    else:
+        example_name = args.example
+        pkg_examples_path = _get_ert3_examples_path()
+        pkg_example_path = pkg_examples_path / example_name
+        wd_example_path = pathlib.Path.cwd() / example_name
+
+        # check that examples folder contains provided 'example_name'
+        if not pkg_example_path.exists():
+            raise ert3.exceptions.IllegalWorkspaceOperation(
+                f"Example {example_name} is not a valid ert3 example."
```
Comment (markusdregi) [inline ert3/console/_console.py:123]: Should we list the available examples? If moved into a function as suggested above, it should be quite easy to append the available examples 👍🏻

File: tests/ert3/console/integration/test_cli.py [hunk 21]
```diff
@@ -68,6 +69,64 @@ def test_cli_init_subfolder(workspace):
...
+
+
+@pytest.mark.requires_ert_storage
+def test_cli_init_example(tmpdir, ert_storage):
+    with tmpdir.as_cwd():
+        args = ["ert3", "init", "--example", "polynomial"]
+        with patch.object(sys, "argv", args):
+            ert3.console.main()
+        polynomial_path = pathlib.Path(tmpdir / "polynomial")
+        if polynomial_path.exists() and polynomial_path.is_dir():
```
Comment (markusdregi) [inline tests/ert3/console/integration/test_cli.py:90]: Why is this test gracefully handing the case where `polynomial_path` does not exist? Should it not fail if it is not present? To me it seems like the `if` should be made into an `assert` and the following `assert` should be moved one indentation line in 🤔

**Review by markusdregi — COMMENTED** [review]
This is looking really good 👏🏻 

Had a few comments which hopefully have small consequences. When those are resolved I'm ready to merge this PR 😎

**Comment by markusdregi** [comment]
Also, I think you should squash all but the first commit...

File: tests/ert3/console/integration/test_cli.py [hunk 22]
```diff
@@ -122,7 +122,7 @@ def test_cli_init_example_inside_example(tmpdir, ert_storage):
         with patch.object(sys, "argv", args):
             with pytest.raises(
                 SystemExit,
-                match="Already inside an ERT workspace",
+                match="Already inside an ERT example workspace.",
```
Comment (markusdregi) [inline tests/ert3/console/integration/test_cli.py:125]: We should not separate between an `example workspace` and a `workspace`. I suggest you adapt the error message and keep the test as is...

File: ert3/console/_console.py [hunk 23]
```diff
@@ -106,6 +103,14 @@ def _get_ert3_examples_path():
     return pkg_examples_path
 
 
+def _get_ert3_example_names(pkg_examples_path):
```
Comment (markusdregi) [inline ert3/console/_console.py:106]: Could even stop taking an argument and just call `_get_ert3_examples_path` itself :)

Comment (kvashchuka) [reply ert3/console/_console.py:106]: sure

File: tests/ert3/console/integration/test_cli.py (hunk 21 above)
Comment (kvashchuka) [reply tests/ert3/console/integration/test_cli.py:90]: good point

File: ci/jenkins/testkomodo.sh [hunk 24]
```diff
@@ -2,7 +2,7 @@
 copy_test_files () {
     cp -r $CI_SOURCE_ROOT/tests $CI_TEST_ROOT/tests
     cp -r $CI_SOURCE_ROOT/test-data $CI_TEST_ROOT/test-data
-    cp -r $CI_SOURCE_ROOT/examples $CI_TEST_ROOT/examples
+    cp -r $CI_SOURCE_ROOT/ert3_examples $CI_TEST_ROOT/ert3_examples
```
Comment (markusdregi) [inline ci/jenkins/testkomodo.sh:5]: This change should be moved to the first commit

**Review by markusdregi — COMMENTED** [review]
Looking good 🚀 One last comment

**Review by markusdregi — APPROVED** [review]
Thank you for your patience and effort 🙏🏻

**Comment by kvashchuka** [comment]
> Thank you for your patience and effort 🙏🏻

Thank you :)
